### ISSUE TITLE
Question about feature, or possible suggestion

### ISSUE THREAD

**Issue by murphman300** [socketio/socket.io-client-swift#1013]
Hello, 

Been using this framework for quite some time now, and have just been wondering is there is a middleware approach to handling incoming SocketClient handlers? What I mean by this is something that resembles Express' middleware handlers, but in this case I'm thinking of having something that can parse the JSON packets that the client recieves.

The reason I'm asking is because I'm taking advantage of Swift 4's Encodable/Decodable protocols to parse Rest response bodies with my own API and I just realized that this would be quite useful for SocketIO-client-swift.

Is there a way to do this with a middleware-like method i'm not finding?

Otherwise, might it be possible to do this by adding an Encodable-type parameter to .on methods? Where if this parameter is provided for the handler, then your client engine would attempt to cast the incoming packet to an array of that Encodable-type? And subsequently, send the error cought by the do block into the client's 'error' handler.

I think this could make Socket.IO even more fluid then it already is. If this is already implemented, let me know!

Cheers

**Comment by nuclearace** [comment]
Yes I thought about adding such a feature. And one does exist for sending `SocketData`. The issue is that Swift's codable doesn't really play nice with unknown payloads with dictionary and arrays. Not saying it's impossible, just that the interface exposed probably isn't going to be very pretty or intuitive.

You're welcome to give it a shot though, I don't have time much these days to work on stuff in-depth.
